File: projects/alibi/engine-python/src/alibi_engine/archive.py
"""The archive: generated testimony, and the baseline retriever.

Everything here is deterministic from the game seed — templates and draws, no
LLM anywhere (the engine rule). The Java engine reproduces this byte for byte,
so the draw order is spec:

    1. shuffle the 16 non-solution elements, take 8  -> the exonerated set
    2. shuffle the witness pool                      -> first 3 lie, rest are honest
    3. one neutral-spot pick per who-dimension exoneration, in build order
    4. per gossip document: one template pick, one suspect pick
    5. one final shuffle over all assembled documents, then ids doc-001..

The truth model is deliberately simple enough to teach:

- An **exoneration** truthfully rules one non-solution element out. Only 8 of
  the 16 get one — the archive alone can never solve the case; the table can.
- A **red herring** uses the *same templates* to "rule out" each solution
  element. Indistinguishable by style, on purpose: the only way to catch one
  is cross-checking its witness.
- Every red herring's witness is undermined by exactly one truthful
  **counter** document — the solvability guarantee of game-rules.md.
- **Gossip** asserts nothing eliminable. It exists so retrieval has noise to
  rank against, and so the fiction breathes.

Which documents lied is revealed only in `game_ended.red_herrings`.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from .case import DISPLAY, WHO, Case, dimension_of
from .rng import Rng

#: Default results per search, and the per-turn allowance lives in GameConfig.
SEARCH_K = 3

WITNESSES: tuple[str, ...] = (
    "head waiter Colin Pereira",
    "security guard Asha Nair",
    "pianist Leo Fernandes",
    "housekeeper Rekha Iyer",
    "bartender Sam Dutta",
    "florist Maria Gomes",
    "doorman Ravi Menon",
    "sous-chef Priya Nayak",
)

#: Locations that are deliberately NOT place elements, so an alibi for a
#: suspect never accidentally says anything about a place.
NEUTRAL_SPOTS: tuple[str, ...] = ("main stage", "reception desk", "front lawn", "buffet line")

#: Why each method was impossible — true of every method except the one used.
METHOD_FACTS: dict[str, str] = {
    "sleight-of-hand": "the display case was fitted with a weight sensor, and it never tripped — nothing was lifted by hand",
    "duplicate-key": "the vault key never left the manager's chain, and the hourly key checks all passed",
    "service-hatch": "the service hatch was bolted and painted shut since the spring renovation",
    "blackout": "the generators kept every light burning all evening; there was no blackout",
    "forged-pass": "every pass scanned that night matched the printed guest register exactly",
}

GOSSIP_TEMPLATES: tuple[str, ...] = (
    "{witness} remarks: {suspect} seemed nervous all evening, checking the time again and again.",
    "{witness} recalls: {suspect} asked twice when the sapphire viewing would end.",
    "{witness} mentions: {suspect} and the auctioneer argued about money earlier in the week.",
    "{witness} says: {suspect} left the centenary toast early, glass still full.",
)


@dataclass(frozen=True)
class Document:
    """One archive entry. Only id/kind/text ever reach the transcript;
    the truth fields exist for game_ended, the eval, and the tests."""

    id: str
    kind: str
    text: str
    asserts_not: str | None  # element this doc claims to rule out, if any
    truthful: bool
    witness: str | None

    def payload(self) -> dict:
        return {"id": self.id, "kind": self.kind, "text": self.text}


def _tokens(text: str) -> list[str]:
    return re.findall(r"[a-z0-9]+", text.lower())


class Archive:
    """The generated corpus plus the baseline retriever."""

    def __init__(self, documents: list[Document]) -> None:
        self.documents = documents
        self._by_id = {d.id: d for d in documents}
        self._doc_tokens = {d.id: _tokens(d.text) for d in documents}

    def get(self, doc_id: str) -> Document:
        return self._by_id[doc_id]

    def red_herrings(self) -> list[str]:
        return [d.id for d in self.documents if not d.truthful and d.asserts_not is not None]

    def search(self, query: str, k: int = SEARCH_K) -> list[Document]:
        """Deterministic keyword retrieval — integers only, so the Java port
        cannot drift. Score = overlap of unique query tokens with the
        document's tokens; ties break toward shorter documents, then id order.
        """
        want = set(_tokens(query))
        scored = []
        for doc in self.documents:
            toks = self._doc_tokens[doc.id]
            score = len(want & set(toks))
            if score > 0:
                scored.append((-score, len(toks), doc.id))
        scored.sort()
        return [self._by_id[doc_id] for _, _, doc_id in scored[:k]]


def _exoneration(element: str, witness: str, spot: str | None) -> tuple[str, str]:
    """Template shared by truths and red herrings — style must not leak truth."""
    dim = dimension_of(element)
    if dim == "who":
        return ("witness_statement",
                f"{witness} states: {DISPLAY[element]} never left the {spot} "
                f"between ten and midnight — half the room can confirm it.")
    if dim == "how":
        return ("forensic_note", f"{witness} confirms: {METHOD_FACTS[element]}.")
    return ("staff_log",
            f"{witness}'s log: {DISPLAY[element]} was locked and under continuous "
            f"watch from nine o'clock; nobody entered.")


def generate(case: Case, rng: Rng) -> Archive:
    non_solution = [d for d in _all_elements() if d not in case.solution.values()]
    exonerated = sorted(rng.sample(non_solution, 8), key=_all_elements().index)

    pool = list(WITNESSES)
    rng.shuffle(pool)
    liars, honest = pool[:3], pool[3:]

    docs: list[tuple[str, str, str | None, bool, str | None]] = []

    for i, element in enumerate(exonerated):
        witness = honest[i % len(honest)]
        spot = NEUTRAL_SPOTS[rng.below(len(NEUTRAL_SPOTS))] if dimension_of(element) == "who" else None
        kind, text = _exoneration(element, witness, spot)
        docs.append((kind, text, element, True, witness))

    for i, dim in enumerate(("who", "how", "where")):
        element = case.solution[dim]
        witness = liars[i]
        spot = NEUTRAL_SPOTS[rng.below(len(NEUTRAL_SPOTS))] if dim == "who" else None
        _, text = _exoneration(element, witness, spot)
        docs.append(("witness_statement", text, element, False, witness))

    for i, liar in enumerate(liars):
        counter = honest[i % len(honest)]
        text = (f"{counter} notes: {liar} left the gala before ten and spent the "
                f"evening in the car park — whatever {liar} says about that night "
                f"is secondhand at best.")
        docs.append(("staff_log", text, None, True, counter))

    for _ in range(6):
        template = GOSSIP_TEMPLATES[rng.below(len(GOSSIP_TEMPLATES))]
        suspect = WHO[rng.below(len(WHO))]
        witness = honest[rng.below(len(honest))]
        text = template.format(witness=witness, suspect=DISPLAY[suspect])
        docs.append(("gossip", text, None, True, witness))

    rng.shuffle(docs)
    return Archive([
        Document(f"doc-{i + 1:03d}", kind, text, asserts, truthful, witness)
        for i, (kind, text, asserts, truthful, witness) in enumerate(docs)
    ])


def _all_elements() -> tuple[str, ...]:
    from .case import ALL_ELEMENTS
    return ALL_ELEMENTS

File: projects/alibi/engine-java/src/main/java/com/llmeval/alibi/engine/Archive.java
package com.llmeval.alibi.engine;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * The archive: generated testimony, and the baseline retriever. Direct port of the Python
 * engine's {@code archive.py} — every template must render <em>byte for byte</em> identically,
 * because the conformance digest covers {@code archive_generated}. A stray comma here fails all
 * twenty vectors, which is the point.
 *
 * <p>Draw order is spec (see the Python module docstring); the retriever is integer-only so the
 * two languages cannot drift on ranking.
 */
public final class Archive {

    public static final int SEARCH_K = 3;

    static final List<String> WITNESSES = List.of(
            "head waiter Colin Pereira",
            "security guard Asha Nair",
            "pianist Leo Fernandes",
            "housekeeper Rekha Iyer",
            "bartender Sam Dutta",
            "florist Maria Gomes",
            "doorman Ravi Menon",
            "sous-chef Priya Nayak");

    static final List<String> NEUTRAL_SPOTS = List.of(
            "main stage", "reception desk", "front lawn", "buffet line");

    static final Map<String, String> METHOD_FACTS = Map.of(
            "sleight-of-hand",
            "the display case was fitted with a weight sensor, and it never tripped — nothing was lifted by hand",
            "duplicate-key",
            "the vault key never left the manager's chain, and the hourly key checks all passed",
            "service-hatch",
            "the service hatch was bolted and painted shut since the spring renovation",
            "blackout",
            "the generators kept every light burning all evening; there was no blackout",
            "forged-pass",
            "every pass scanned that night matched the printed guest register exactly");

    static final List<String> GOSSIP_TEMPLATES = List.of(
            "{witness} remarks: {suspect} seemed nervous all evening, checking the time again and again.",
            "{witness} recalls: {suspect} asked twice when the sapphire viewing would end.",
            "{witness} mentions: {suspect} and the auctioneer argued about money earlier in the week.",
            "{witness} says: {suspect} left the centenary toast early, glass still full.");

    private static final Pattern TOKEN = Pattern.compile("[a-z0-9]+");

    /** One archive entry. Only id/kind/text reach the transcript. */
    public record Document(String id, String kind, String text,
                           String assertsNot, boolean truthful, String witness) {

        public Map<String, Object> payload() {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("id", id);
            map.put("kind", kind);
            map.put("text", text);
            return map;
        }
    }

    private final List<Document> documents;
    private final Map<String, List<String>> docTokens = new LinkedHashMap<>();

    Archive(List<Document> documents) {
        this.documents = documents;
        for (Document doc : documents) {
            docTokens.put(doc.id(), tokens(doc.text()));
        }
    }

    public List<Document> documents() {
        return documents;
    }

    public List<String> redHerrings() {
        List<String> ids = new ArrayList<>();
        for (Document doc : documents) {
            if (!doc.truthful() && doc.assertsNot() != null) {
                ids.add(doc.id());
            }
        }
        return ids;
    }

    /** Deterministic keyword retrieval — integers only, mirroring the Python spec exactly. */
    public List<Document> search(String query, int k) {
        Set<String> want = new HashSet<>(tokens(query));
        record Hit(int score, int length, String id) {}
        List<Hit> hits = new ArrayList<>();
        for (Document doc : documents) {
            List<String> toks = docTokens.get(doc.id());
            Set<String> overlap = new HashSet<>(want);
            overlap.retainAll(new HashSet<>(toks));
            if (!overlap.isEmpty()) {
                hits.add(new Hit(overlap.size(), toks.size(), doc.id()));
            }
        }
        hits.sort(Comparator.comparingInt((Hit h) -> -h.score())
                .thenComparingInt(Hit::length)
                .thenComparing(Hit::id));
        List<Document> results = new ArrayList<>();
        for (Hit hit : hits.subList(0, Math.min(k, hits.size()))) {
            for (Document doc : documents) {
                if (doc.id().equals(hit.id())) {
                    results.add(doc);
                    break;
                }
            }
        }
        return results;
    }

    public List<Document> search(String query) {
        return search(query, SEARCH_K);
    }

    static List<String> tokens(String text) {
        List<String> out = new ArrayList<>();
        Matcher m = TOKEN.matcher(text.toLowerCase(Locale.ROOT));
        while (m.find()) {
            out.add(m.group());
        }
        return out;
    }

    // -- generation ------------------------------------------------------

    private record Row(String kind, String text, String assertsNot, boolean truthful, String witness) {}

    /** Kind and text for one exoneration — the template shared by truths and red herrings. */
    private static Row exoneration(String element, String witness, String spot, boolean truthful,
                                   boolean forceWitnessStatement) {
        String dim = CaseModel.dimensionOf(element);
        String kind;
        String text;
        if (dim.equals("who")) {
            kind = "witness_statement";
            text = witness + " states: " + CaseModel.DISPLAY.get(element)
                    + " never left the " + spot
                    + " between ten and midnight — half the room can confirm it.";
        } else if (dim.equals("how")) {
            kind = "forensic_note";
            text = witness + " confirms: " + METHOD_FACTS.get(element) + ".";
        } else {
            kind = "staff_log";
            text = witness + "'s log: " + CaseModel.DISPLAY.get(element)
                    + " was locked and under continuous watch from nine o'clock; nobody entered.";
        }
        if (forceWitnessStatement) {
            kind = "witness_statement";
        }
        return new Row(kind, text, element, truthful, witness);
    }

    public static Archive generate(CaseModel caseModel, Rng rng) {
        List<String> nonSolution = new ArrayList<>();
        for (String element : CaseModel.ALL_ELEMENTS) {
            if (!caseModel.solution().containsValue(element)) {
                nonSolution.add(element);
            }
        }
        List<String> exonerated = rng.sample(nonSolution, 8);
        exonerated.sort(Comparator.comparingInt(CaseModel.ALL_ELEMENTS::indexOf));

        List<String> pool = new ArrayList<>(WITNESSES);
        rng.shuffle(pool);
        List<String> liars = List.copyOf(pool.subList(0, 3));
        List<String> honest = List.copyOf(pool.subList(3, pool.size()));

        List<Row> rows = new ArrayList<>();

        for (int i = 0; i < exonerated.size(); i++) {
            String element = exonerated.get(i);
            String witness = honest.get(i % honest.size());
            String spot = CaseModel.dimensionOf(element).equals("who")
                    ? NEUTRAL_SPOTS.get(rng.below(NEUTRAL_SPOTS.size())) : null;
            rows.add(exoneration(element, witness, spot, true, false));
        }

        List<String> dims = List.of("who", "how", "where");
        for (int i = 0; i < dims.size(); i++) {
            String element = caseModel.solution().get(dims.get(i));
            String witness = liars.get(i);
            String spot = dims.get(i).equals("who")
                    ? NEUTRAL_SPOTS.get(rng.below(NEUTRAL_SPOTS.size())) : null;
            rows.add(exoneration(element, witness, spot, false, true));
        }

        for (int i = 0; i < liars.size(); i++) {
            String liar = liars.get(i);
            String counter = honest.get(i % honest.size());
            String text = counter + " notes: " + liar
                    + " left the gala before ten and spent the evening in the car park — whatever "
                    + liar + " says about that night is secondhand at best.";
            rows.add(new Row("staff_log", text, null, true, counter));
        }

        for (int i = 0; i < 6; i++) {
            String template = GOSSIP_TEMPLATES.get(rng.below(GOSSIP_TEMPLATES.size()));
            String suspect = CaseModel.WHO.get(rng.below(CaseModel.WHO.size()));
            String witness = honest.get(rng.below(honest.size()));
            String text = template
                    .replace("{witness}", witness)
                    .replace("{suspect}", CaseModel.DISPLAY.get(suspect));
            rows.add(new Row("gossip", text, null, true, witness));
        }

        rng.shuffle(rows);

        List<Document> documents = new ArrayList<>();
        for (int i = 0; i < rows.size(); i++) {
            Row row = rows.get(i);
            documents.add(new Document(String.format("doc-%03d", i + 1), row.kind(), row.text(),
                    row.assertsNot(), row.truthful(), row.witness()));
        }
        // The ids in redHerrings() must come out ascending; assembly order guarantees it.
        Set<String> unique = new LinkedHashSet<>();
        for (Document doc : documents) {
            unique.add(doc.id());
        }
        if (unique.size() != documents.size()) {
            throw new IllegalStateException("duplicate document ids");
        }
        return new Archive(documents);
    }
}
